app.js

```diff
@@ -137,38 +138,167 @@ class Products {
+    closeCartBtn.addEventListener("click", this.hideCart);
+  }
+  populateCart(cart) {
+    cart.forEach((item) => this.addCartItems(item));
+  }
+  //to close cart
+  hideCart() {
+    cartOverlay.classList.remove("transparentBcg");
+    cartDom.classList.remove("showCart");
+  }
+
+  //A method for the cart logic
+  cartLogic() {
+    cartContent.addEventListener("click", (event) => {
+      //For the remove button in the cart
+      if (event.target.classList.contains("remove-item")) {
+        let removeItem = event.target;
+        let id = removeItem.dataset.id;
+        cartContent.removeChild(removeItem.parentElement.parentElement);
+        this.removeItem(id);
+      }
+      //for the increament of the variant
+      else if (event.target.classList.contains("fa-chevron-up")) {
+        let increaseNumber = event.target;
+        let id = increaseNumber.dataset.id;
+        let tempItem = cart.find((item) => item.id === id);
+        tempItem.amount = tempItem.amount + 1;
+        Storage.saveCart(cart);
+        this.setCartValues(cart);
+        increaseNumber.nextElementSibling.innerText = tempItem.amount;
+      }
+      //for the reduction of the variant
+      else if (event.target.classList.contains("fa-chevron-down")) {
+        let lowerAmount = event.target;
+        let id = lowerAmount.dataset.id;
+        let tempItem = cart.find((item) => item.id === id);
+        tempItem.amount = tempItem.amount - 1;
+        if (tempItem.amount > 0) {
+          Storage.saveCart(cart);
+          this.setCartValues(cart);
+          lowerAmount.previousElementSibling.innerText = tempItem.amount;
+        } else {
+          cartContent.removeChild(lowerAmount.parentElement.parentElement);
+          this.removeItem(id);
+        }
+      }
+    });
+    //For the clear cart  button event
+    //clearCartBtn.addEventListener('click',() => {
+    //this.clearCart();
+    //});
+  }
+  //For the Clear cart button Functionalities.
+  //clearCart() {
+  //let cartItems = cart.map(item => item.id);
+  //cartItems.forEach(id => this.removeItem(id))
+
+  //while(cartContent.children.length >0) {
+  //  cartContent.removeChild(cartContent.children[0])
+  // }
+  // this.hideCart();
+  //}
+
+  //to remove the items from the cart doing that with their ID.
+  removeItem(id) {
+    cart = cart.filter((item) => item.id !== id);
+    this.setCartValues(cart);
+    Storage.saveCart(cart);
+    let button = this.getSingleButton(id);
+    button.disabled = false;
+    button.innerHTML = `<i class ="fas fa-shopping-cart">add to cart</i>`;
+  }
+  getSingleButton(id) {
+    return buttonsDOM.find((button) => button.dataset.id === id);
+  }
+}
+
+//To store it in the local storage.
+
+class Storage {
+  static saveProducts(products) {
+    localStorage.setItem("products", JSON.stringify(products));
+  }
+  static getProduct(id) {
+    let products = JSON.parse(localStorage.getItem("products"));
+    return products.find((product) => product.id === id);
```

Comment:
this operation is expensive, what happens if you are to get a single product from 1000 product

Reply:
Expensive? I really don't understand.

Reply:
So you mean I should map through it instead ?

Reply:
what i mean here is the complexity of this function is linear O(N)

Reply:
all you need to convert O(N) -> O(1) (constant)